### PR TITLE
Adds handler for kebab-case event for v-bind:sync; fix #6428

### LINKED ISSUE #6428: `v-bind.sync` listens for camelCase event
### What problem does this feature solve?
This markup:

```html
<component :max-items.sync="showItems"></component>
```

Currently compiles to:

```javascript
function render() {
    with(this) {
        return _c('component', {
            attrs: {
                "max-items": showItems
            },
            on: {
                "update:maxItems": function($event) {
                    showItems = $event
                }
            }
        })
    }
}
```

I've always written event handlers in kebab-case. It surprised me that the compiler created an event handler which is in camelCase.

### What does the proposed API look like?
The compiler should output `.sync` event handlers in camelCase:

```javascript
            on: {
                "update:max-items": function($event) {
                    showItems = $event
                }
            }
```

### PR DESCRIPTION
**Pull request by shortdiv**
**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [x] Feature

**Does this PR introduce a breaking change?** (check one)

- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

**Other information:**

Addresses #6428 and https://github.com/vuejs/vuejs.org/issues/1648

### CODE REVIEW

**Comment by chrisvfritz** [comment]
@yyx990803 Is there anything more that should be one on this? Being able to universally recommend kebab-case event names will simplify a couple sections of the docs and allow us to add [this rule](https://github.com/vuejs/eslint-plugin-vue/issues/356) to the ESLint plugin. 🙂

File: src/compiler/parser/index.js [hunk 1]
```diff
@@ -538,10 +538,16 @@ function processAttrs (el) {
             name = camelize(name)
           }
           if (modifiers.sync) {
+            syncGen = genAssignmentCode(value, `$event`)
+            addHandler(
+              el,
+              `update:${kebabize(name)}`,
```
Comment (sodatea) [inline src/compiler/parser/index.js]: The `hyphenate` function might be a better fit here.

Comment (chrisvfritz) [reply src/compiler/parser/index.js]: Good point! 🙂 I forgot that one existed. @shortdiv, can you update to use `hyphenate` instead?

File: src/shared/util.js [hunk 2]
```diff
@@ -158,6 +158,14 @@ export const camelize = cached((str: string): string => {
...
 })
 
+/**
+ * Kebabize a camelCased string
+ */
+const kebabizeRE = /([a-z])([A-Z])/g
+export const kebabize = cached((str: string): string => {
+  return str.replace(kebabizeRE, (m, p1, p2) => `${p1}-${p2}`).toLowerCase()
+})
+
```
Comment (chrisvfritz) [inline src/shared/util.js]: Since we're no longer using this, can you delete it please?

Comment (shortdiv) [reply src/shared/util.js]: You got it!

**Comment by yyx990803** [comment]
// TODO merge #8842 with this one together

File: src/shared/util.js [hunk 3]
```diff
@@ -158,6 +158,14 @@ export const camelize = cached((str: string): string => {
   return str.replace(camelizeRE, (_, c) => c ? c.toUpperCase() : '')
 })
 
+/**
+ * Kebabize a camelCased string
+ */
+const kebabizeRE = /([a-z])([A-Z])/g
+export const kebabize = cached((str: string): string => {
```
Comment (Fl0pZz) [inline src/shared/util.js]: Please explain why the cached function is used

**Comment by Justineo** [comment]
> // TODO revert changes in #8845 that will become unnecessary after merging this

Only [this line](https://github.com/vuejs/vue/pull/8845#discussion-diff-220505690L51) can be reverted. But I think it's okay to keep it as is.

**Comment by chrisvfritz** [comment]
Thinking about this a little more, I wonder if `addHandler` itself should actually create both camelCase and kebab-case listeners for every event. This would fix cases where a library is emitting a camelCased event that can't be listened to in HTML (even though we warn against it in the docs) and also solve a [JSX issue](https://github.com/vuejs/babel-plugin-transform-vue-jsx/issues/164).

**Comment by yyx990803** [comment]
@chrisvfritz that would significantly bloat the compiled output, not sure if it's worth it, especially when this is not an issue when using string templates.

**Pull request merged**